absl/types/any.h

```diff
@@ -103,14 +111,10 @@ namespace any_internal {
 // 32-bit integer. While a client should never do that it SHOULD still be safe,
 // assuming the BSS segment doesn't span more than 4GiB.
 template<typename Type>
-inline size_t FastTypeId() {
+constexpr inline size_t FastTypeId() {
   static_assert(sizeof(char*) <= sizeof(size_t),
                 "ptr size too large for size_t");
-
-  // This static variable isn't actually used, only its address, so there are
-  // no concurrency issues.
-  static char dummy_var;
-  return reinterpret_cast<size_t>(&dummy_var);
+  return reinterpret_cast<size_t>(&TypeTag<Type>::dummy_var);
```

Comment:
Actually, backtracking a bit -- this will not make the call constexpr because of the reinterpret_cast. What is the precise intent here in making it constexpr, given no constexpr uses? We can make this constexpr if we change FastTypeId to not return a size_t, but rather return the pointer directly.

Reply:
I would expect at the very least this will get rid of static initialisation in the function scope - which compiler needs to enforce synchronization of (as per section 6.7). I imagine under any sane optimizer this synchronization would be removed, but trying to make optimizer work easier here and the intent more obvious - both to compiler and to reader.

Yes it is slightly embarrassing that this function should not be allowed to be `constexpr` by a conforming compiler. If returned value was "const void* " or perhaps "const char *", that would not have been a problem. I would even defer the cast to size_t (or perhaps more fittingly, `uintptr_t`) until the body of `GetObjTypeId()` where we have conditional expression on runtime condition anyway.

Reply:
Could you update the return to simply be `const void*` instead, and remove the `reinterpret_cast` as a part of this change? This would require cascading changes to the return type in `IdForType()` and `GetObjTypeId()` as well. Thanks.

Reply:
Thanks Matt, will do.